modules/040-node-manager/images/bashible-apiserver/pkg/template/steps_storage.go

```diff
@@ -17,48 +17,195 @@ limitations under the License.
+	"sync"
+	"time"
+
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/dynamic/dynamicinformer"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog/v2"
 )
 
 // NameMapper maps the name of resource to the secretKey of a template
 type NameMapper func(name string) (string, error)
 
-type config struct {
-	rootDir  string
-	provider string
-	target   string
-}
-
 // StepsStorage is the storage if bashible steps for a particular steps target
 type StepsStorage struct {
-	config config
+	rootDir string
+
+	m sync.RWMutex
+	// cache all system scripts with lazy load
+	systemScripts map[string]map[string][]byte
+
+	nodeGroupConfigurations map[string][]*nodeConfigurationScript
+
+	configurationsChanged chan struct{}
+	emitter               changesEmitter
 }
 
-// NewStepsStorage creates StepsStorage for target and cloud provider.
-//
-// provider = aws | gcp | openstack | ...
-// target   = all | node-group
-func NewStepsStorage(rootDir, provider, target string) *StepsStorage {
-	config := config{rootDir, provider, target}
-	return &StepsStorage{config}
+type changesEmitter struct {
+	sync.Mutex
+	count int
 }
 
-func (s *StepsStorage) Render(templateContext map[string]interface{}) (map[string]string, error) {
-	bundle, ok := templateContext["bundle"].(string)
-	if !ok {
-		panic("expected string in templateContext[\"bundle\"]")
+// NewStepsStorage creates StepsStorage for target and cloud provider.
+func NewStepsStorage(ctx context.Context, rootDir string, ngConfigFactory dynamicinformer.DynamicSharedInformerFactory) *StepsStorage {
+	ss := &StepsStorage{
+		rootDir:                 rootDir,
+		systemScripts:           make(map[string]map[string][]byte),
+		nodeGroupConfigurations: make(map[string][]*nodeConfigurationScript),
+		configurationsChanged:   make(chan struct{}, 1),
 	}
 
-	templates, err := s.readBundleTemplates(bundle)
+	ss.subscribeOnCRD(ctx, ngConfigFactory)
+	return ss
+}
+
+func (s *StepsStorage) Render(target, bundle, provider string, templateContext map[string]interface{}, ng ...string) (map[string]string, error) {
+	steps, err := s.renderSystemScripts(target, bundle, provider, templateContext)
 	if err != nil {
 		return nil, err
 	}
 
-	steps := map[string]string{}
+	if len(ng) > 0 {
+		userConfigurations, err := s.renderNodeGroupConfigurations(bundle, ng[0], templateContext)
+		if err != nil {
+			klog.Errorf("Render user NodeGroupConfigurations failed: %s", err)
+			return steps, nil
+		}
+
+		for k, v := range userConfigurations {
+			if _, ok := steps[k]; ok {
+				klog.Errorf("NodeGroupConfigurations conflicts with system script: %s", k)
+				continue
+			}
+			steps[k] = v
+		}
+	}
+
+	return steps, nil
+}
+
+func (s *StepsStorage) OnNodeConfigurationsChanged() chan struct{} {
+	return s.configurationsChanged
+}
+
+func (s *StepsStorage) runBufferedEmitter() chan struct{} {
+	for {
+		time.Sleep(500 * time.Millisecond)
```

Comment:
Didn't get the purpose of this sleep.

Reply:
To not emit a few time on batch updates, for example on start or on applying a few configurations. Because every configuration emitted will rerun content rendering

Reply:
Ok, it makes sense. Could you please comment in the code on why we need this sleep?

Reply:
Done